### PR TITLE
More corner cases with iptables proxy

### PR DESCRIPTION
**Pull request by thockin**
Node ports
Set sysctls
Cleanup rules code
Faster sync at startup
Better logging

DNAT back to self doesn't quite work yet because we need to enable hairpin mode on the bridge, and that has to be done interface-by-interface (sigh).  Will keep hunting for a better answer there.

This needs SERIOUS e2e testing.  I have a very manual battery of tests I have been running, which is a good starting place.

@BenTheElder @ArtfulCoder

### CODE REVIEW

**Comment by BenTheElder** [comment]
Read through current last commit (https://github.com/thockin/kubernetes/commit/23a3ab8355eb48317dc2860a394ba76b5001a473).
LGTM

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -434,126 +510,213 @@ func (proxier *Proxier) syncProxyRules() error {
...
+			args := []string{
+				"-A", string(svcChain),
+				"-m", "comment", "--comment", name.String(),
 			}
+			if i < (n - 1) {
+				// Each rule is a probabilistic match.
+				args = append(args,
+					"-m", "statistic",
+					"--mode", "random",
+					"--probability", fmt.Sprintf("%f", 1.0/float64(n-i)))
```
Comment (ArtfulCoder) [inline pkg/proxy/iptables/proxier.go]: (discussed offline)
we should probably restrict decimal length

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -122,10 +160,24 @@ var _ proxy.ProxyProvider = &Proxier{}
 // An error will be returned if iptables fails to update or acquire the initial lock.
 // Once a proxier is created, it will keep iptables up to date in the background and
 // will not terminate if a particular iptables call fails.
-func NewProxier(ipt utiliptables.Interface, syncPeriod time.Duration) (*Proxier, error) {
+func NewProxier(ipt utiliptables.Interface, exec utilexec.Interface, syncPeriod time.Duration) (*Proxier, error) {
 	glog.V(2).Info("Tearing down userspace rules. Errors here are acceptable.")
 	// remove iptables rules/chains from the userspace Proxier
 	tearDownUserspaceIptables(ipt)
```
Comment (ArtfulCoder) [inline pkg/proxy/iptables/proxier.go]: (discussed offline) we can push the tearDownUserspaceIptables to a later point in the execution flow

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Done in next push

**Comment by thockin** [comment]
Issues fixed, new push is up.  This needs e2e before it can be activated but I'd like to merge it ASAP.

**Comment by thockin** [comment]
@k8s-bot test this please

**Comment by ArtfulCoder** [comment]
LGTM

**Comment by BenTheElder** [comment]
Whoo!

**Comment by BenTheElder** [comment]
So we still need hairpin mode, anything else?

**Comment by thockin** [comment]
hairpin mode (hairy)
maybe load more kernel modules (just need to understand why)
maybe add "always masq" mode as a flag (easy)
hold node ports open (in progress)

Not sure what else.  If you have extra bandwidth, you could do the "always
masq" mode.   Is there more to do?

**Comment by BenTheElder** [comment]
I'll take a look at the always masq flag this afternoon. As far as I know
that sounds like it, except maybe this:
https://github.com/mesosphere/kubernetes-mesos/issues/353#issuecomment-127382832

**Comment by thockin** [comment]
yeah, maybe.  It's P2 or P3 to me.  The other thing we could use is better
e2e testing of this new mode.  That's a non-trivial undertaking.  I'm happy
to consume more of your time, but I know you're back in school...

**Comment by BenTheElder** [comment]
Yeah, I'll see if I can get a bit more done early this week before i get
_really_ busy and GSoC ends. So far it's mostly syllabi and realizing I
need to review a few things. Should have time at least to knock out always
masq mode.

**Pull request merged**